repository/src/fetch.rs

```diff
@@ -0,0 +1,269 @@
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
+
+        // Skip if the branch is `fp`.
+        if branch_name == FP_BRANCH_NAME {
+            continue;
+        }
+
+        // Check if the branch is already in the local repo.
+        if local_branches
+            .iter()
+            .any(|(_, commit_hash_)| *commit_hash_ == commit_hash)
+        {
+            info!("skip {}: alreadd tracked by local repo", branch_displayed);
+            continue;
+        }
+
+        // Check if the branch is rebased on the finalized branch.
+        if last_finalized_commit_hash
+            != this
+                .raw
+                .find_merge_base(last_finalized_commit_hash, commit_hash)
+                .await?
+        {
+            info!("remote tracking branch outdated: {}", branch_displayed);
+            continue;
+        }
+
+        // Reads the commits in the branch.
+        let commits = match read_commits(this, last_finalized_commit_hash, commit_hash).await {
+            Ok(x) => x,
+            Err(CommitError::Commit(error, commit)) => {
+                warn!("failed to parse commit {}: {}", commit, error);
+                continue;
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        // Verify all the incoming commits
+        let mut csv = CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+            .expect("finalized branch is not accepted by CSV");
+        for (new_commit, new_commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(new_commit) {
+                warn!(
+                    "commit verification failed for branch {}: {} at {}",
+                    branch_displayed, e, new_commit_hash
+                );
+                continue 'branch_loop;
+            }
+        }
+
+        next_step_branches.push((
+            remote_name,
+            branch_name,
+            commit_hash,
+            commits.last().unwrap().0.clone(),
+        ));
+
+        // Store the block commits (including the currently finalized one)
+        let block_commit_hashes = std::iter::once(last_finalized_commit_hash)
+            .chain(
+                commits
+                    .iter()
+                    .filter_map(|(commit, hash)| matches!(commit, Commit::Block(_)).then(|| *hash)),
+            )
+            .collect::<Vec<_>>();
+
+        // Try to finalized the last block with the known finalization proofs.
+        let mut hit = false;
+        for proof in &finalization_proofs {
+            if csv.verify_last_header_finalization(&proof.proof).is_ok() {
+                last_finalization_proof = proof.clone();
+                last_finalized_commit_hash = *block_commit_hashes.last().unwrap();
+                hit = true;
+            }
+        }
+
+        // If failed, finalize the last-last block.
+        if !hit && block_commit_hashes.len() > 2 {
```

Comment:
Can we pass the len =1, len = 0 case?

Reply:
In such cases, it means that there is no second to last block commit

Reply:
@YoonhoKim0527 check and resolve please